tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3059,5 +3106,169 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
   return Status::OK();
 }
 
+// Overall, emit code for slices based on the below structure. A `slice_guard_i`
+// and a `slice_i` are generated for each ROOT slice. `slice_guard_i` computes
+// the guarding condition to decide whether it should jump into `slice_i`
+// for writing to the slice output or continue to next `slice_guard_{i+1}`.
+//
+// init_block:
+//   Compute values of slice input operands
+//   Br slice_guard_0
+//
+// slice_guard_0:
```

Comment:
Sorry for nitpicking, I'm still trying to understand whether we actually need non-structured control flow here.

Is the code snippet below equivalent to

```
if (guarding_cond0) {
  // Write to output of slice0
}
if (guarding_cond1) {
  // Write to output  of slice 1
}
...

?
```

Reply:
Gave it some more thoughts now. Yes, it can be written in a structured control flow. I did not pay too much attention to it as I initially use the concat codegen as my reference point. If this really bothers you, I can try to re-write it into structured control flow. Please let me know what you feel.

Reply:
I do not want to block your work on this, but it seems that structured flow would be easier to read and maintain. Do you want to try it and see whether it is indeed the case? I'm having a hard time following what exactly the emitter is doing due to all the stateful blocks.

Reply:
No problem. Let's me give it a try. I think it will work.

Reply:
Done the rewriting. The code is indeed much cleaner. Thanks for asking the good question.